File: windows/tests/Elevate.Core.Tests/TokenCacheHintTests.cs
using Elevate.Core.Coordination;
using Elevate.Core.Models;
using Elevate.Core.Support;
using FluentAssertions;

namespace Elevate.Core.Tests;

public class TokenCacheHintTests
{
    private static readonly DateTimeOffset Now = DateTimeOffset.FromUnixTimeSeconds(1_000_000);

    private static RoleKey Entra(string identity = "i") => new(identity, "t", new EntraDirectoryScope("r", "/"));

    private static RoleKey Azure(string identity = "i") => new(identity, "t", new AzureResourceScope("/subscriptions/s1", "owner"));

    private static RoleKey Group(string identity = "i") => new(identity, "t", new GroupScope("g1", GroupAccess.Member));

    private static ActivationOutcome Activated(RoleKey key) =>
        new(key, new ActivationResult.Activated(new ActiveAssignment(key, "a", Now, Now.AddHours(1), AssignmentStatus.Active)));

    private static ActivationOutcome Pending(RoleKey key) =>
        new(key, new ActivationResult.PendingApproval(new ActiveAssignment(key, "p", Now, null, AssignmentStatus.PendingApproval)));

    /// <summary>Every account signed in through the Azure CLI app, so the tool cache is shared.</summary>
    private static SignInMethod? Cli(string _) => SignInMethod.AzureCLI;

    [Fact]
    public void AzureAndGroupActivationsQualifyEntraDoesNot()
    {
        TokenCacheHint.Qualifies(Azure()).Should().BeTrue();
        TokenCacheHint.Qualifies(Group()).Should().BeTrue();
        TokenCacheHint.Qualifies(Entra()).Should().BeFalse();
    }

    [Fact]
    public void AffectedAccountsListsEachAccountOnceForActivatedAzureOrGroupRoles()
    {
        var outcomes = new[]
        {
            Activated(Entra("a")),
            Activated(Azure("a")),
            Activated(Group("a")),
            Activated(Group("b")),
            Pending(Azure("c")),
            new ActivationOutcome(Azure("d"), new ActivationResult.Failed(new PimException(PimErrorKind.Forbidden, "no"))),
        };

        TokenCacheHint.AffectedAccounts(outcomes, Cli).Should().Equal("a", "b");
    }

    [Fact]
    public void OnlyAccountsSharingTheToolTokenCacheAreAffected()
    {
        var methods = new Dictionary<string, SignInMethod>
        {
            ["cli"] = SignInMethod.AzureCLI,
            ["ps"] = SignInMethod.AzurePowerShell,
            ["app"] = SignInMethod.OwnApp,
            ["custom"] = SignInMethod.Custom("11111111-1111-1111-1111-111111111111"),
        };
        var outcomes = new[] { Activated(Azure("app")), Activated(Azure("cli")), Activated(Group("custom")), Activated(Group("ps")), Activated(Azure("gone")) };

        // An app registration has its own cache: Elevate cannot tell whether the tools were ever used as that account.
        TokenCacheHint.AffectedAccounts(outcomes, id => methods.GetValueOrDefault(id)).Should().Equal("cli", "ps");
        SignInMethod.AzureCLI.SharesToolTokenCache.Should().BeTrue();
        SignInMethod.AzurePowerShell.SharesToolTokenCache.Should().BeTrue();
        SignInMethod.OwnApp.SharesToolTokenCache.Should().BeFalse();
        methods["custom"].SharesToolTokenCache.Should().BeFalse();
    }

    [Fact]
    public void NothingQualifyingMeansNoAccounts()
    {
        TokenCacheHint.AffectedAccounts([Activated(Entra()), Pending(Group())], Cli).Should().BeEmpty();
        TokenCacheHint.AffectedAccounts([], Cli).Should().BeEmpty();
    }

    [Fact]
    public void WordingNamesTheAccountAndTheExactCommands()
    {
        TokenCacheHint.Message("alex@contoso.com").Should().Contain("alex@contoso.com").And.Contain("Azure CLI");
        TokenCacheHint.Advice.Should().Contain("az login").And.Contain("kubelogin remove-tokens");
        TokenCacheHint.Advice.Should().NotContain("--force-refresh", "az has no such flag; signing in again is the only way");
    }
}

File: macos/Tests/ElevateCoreTests/TokenCacheHintTests.swift
import Testing
import Foundation
@testable import ElevateCore

@Suite struct TokenCacheHintTests {
    let now = Date(timeIntervalSince1970: 1_000_000)

    func entra(_ identity: String = "i") -> RoleKey { RoleKey(identityId: identity, tenantId: "t", scope: .entraDirectory(roleDefinitionId: "r", directoryScopeId: "/")) }
    func azure(_ identity: String = "i") -> RoleKey { RoleKey(identityId: identity, tenantId: "t", scope: .azureResource(scope: "/subscriptions/s1", roleDefinitionId: "owner")) }
    func group(_ identity: String = "i") -> RoleKey { RoleKey(identityId: identity, tenantId: "t", scope: .group(groupId: "g1", accessId: .member)) }

    func activated(_ key: RoleKey) -> ActivationOutcome {
        ActivationOutcome(roleKey: key, result: .activated(ActiveAssignment(roleKey: key, assignmentId: "a", startDateTime: now,
                                                                             endDateTime: now.addingTimeInterval(3600), status: .active)))
    }
    func pending(_ key: RoleKey) -> ActivationOutcome {
        ActivationOutcome(roleKey: key, result: .pendingApproval(ActiveAssignment(roleKey: key, assignmentId: "p", startDateTime: now,
                                                                                   endDateTime: nil, status: .pendingApproval)))
    }

    @Test func azureAndGroupActivationsQualifyEntraDoesNot() {
        #expect(TokenCacheHint.qualifies(azure()))
        #expect(TokenCacheHint.qualifies(group()))
        #expect(!TokenCacheHint.qualifies(entra()))
    }

    /// Every account signed in through the Azure CLI app, so the tool cache is shared.
    func cli(_: String) -> SignInMethod? { .azureCLI }

    @Test func affectedAccountsListsEachAccountOnceForActivatedAzureOrGroupRoles() {
        let outcomes = [activated(entra("a")), activated(azure("a")), activated(group("a")), activated(group("b")),
                        pending(azure("c")), ActivationOutcome(roleKey: azure("d"), result: .failed(.forbidden("no")))]
        #expect(TokenCacheHint.affectedAccounts(outcomes, method: cli) == ["a", "b"])
    }

    @Test func onlyAccountsSharingTheToolTokenCacheAreAffected() {
        let methods: [String: SignInMethod] = ["cli": .azureCLI, "ps": .azurePowerShell, "app": .ownApp,
                                               "custom": .custom(clientId: "11111111-1111-1111-1111-111111111111")]
        let outcomes = [activated(azure("app")), activated(azure("cli")), activated(group("custom")), activated(group("ps")), activated(azure("gone"))]
        // An app registration has its own cache: Elevate cannot tell whether the tools were ever used as that account.
        #expect(TokenCacheHint.affectedAccounts(outcomes) { methods[$0] } == ["cli", "ps"])
        #expect(SignInMethod.azureCLI.sharesToolTokenCache)
        #expect(SignInMethod.azurePowerShell.sharesToolTokenCache)
        #expect(!SignInMethod.ownApp.sharesToolTokenCache)
        #expect(!methods["custom"]!.sharesToolTokenCache)
    }

    @Test func nothingQualifyingMeansNoAccounts() {
        #expect(TokenCacheHint.affectedAccounts([activated(entra()), pending(group())], method: cli).isEmpty)
        #expect(TokenCacheHint.affectedAccounts([], method: cli).isEmpty)
    }

    @Test func wordingNamesTheAccountAndTheExactCommands() {
        #expect(TokenCacheHint.message(account: "alex@contoso.com").contains("alex@contoso.com"))
        #expect(TokenCacheHint.advice.contains("az login"))
        #expect(TokenCacheHint.advice.contains("kubelogin remove-tokens"))
        // az has no flag that forces a fresh token; signing in again is the only way.
        #expect(!TokenCacheHint.advice.contains("--force-refresh"))
    }
}
